crates/modelardb_server/src/storage/types.rs

```diff
@@ -0,0 +1,291 @@
+ * limitations under the License.
+ */
+
+//! The minor types used throughout the [`StorageEngine`](crate::storage::StorageEngine).
+
+use std::fmt;
+use std::mem;
+use std::sync::RwLock;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use datafusion::arrow::array::UInt32Array;
+use modelardb_common::types::{Timestamp, TimestampArray};
+use ringbuf::{HeapRb, Rb};
+
+/// Resizeable pool of memory for tracking and limiting the amount of memory used by the
+/// [`StorageEngine`](crate::storage::StorageEngine).
+pub(super) struct MemoryPool {
+    /// How many bytes of memory that are left for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata.
+    remaining_uncompressed_memory_in_bytes: RwLock<isize>,
+    /// How many bytes of memory that are left for storing
+    /// [`CompressedDataBuffers`](crate::storage::compressed_data_buffer::CompressedDataBuffer). A
+    /// signed integer is used since compressed data is inserted and then the remaining bytes are
+    /// checked. This means that the remaining bytes can briefly be negative until compressed data
+    /// is saved to disk and if the amount of memory for uncompressed data is reduced.
+    remaining_compressed_memory_in_bytes: RwLock<isize>,
+}
+
+/// The pool of memory the
+/// [`UncompressedDataManager`](crate::storage::uncompressed_data_manager::UncompressedDataManager)
+/// can use for uncompressed data and the
+/// [`CompressedDataManager`](crate::storage::CompressedDataManager) can use for compressed data.
+impl MemoryPool {
+    /// Create a new [`MemoryPool`] with at most [`u64::MAX`] bytes of memory for uncompressed data
+    /// and at most [`i64::MAX`] bytes of memory for compressed data.
+    pub(super) fn new(
+        uncompressed_memory_in_bytes: usize,
+        compressed_memory_in_bytes: usize,
+    ) -> Self {
+        // unwrap() is safe as i64::MAX is 8192 PiB and the value is from ConfigurationManager.
+        Self {
+            remaining_uncompressed_memory_in_bytes: RwLock::new(
+                uncompressed_memory_in_bytes.try_into().unwrap(),
+            ),
+            remaining_compressed_memory_in_bytes: RwLock::new(
+                compressed_memory_in_bytes.try_into().unwrap(),
+            ),
+        }
+    }
+
+    /// Change the amount of memory available for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata by `size_in_bytes`.
+    pub(super) fn update_uncompressed_memory(&self, size_in_bytes: isize) {
```

Comment:
Maybe we should try to be consistent in the naming of methods that change a value like this. In another similar method (`adjust_uncompressed_remaining_memory_in_bytes`) we use `adjust` instead of `update`. I am not sure which is best, but we should probably use the same naming. 

If it is changed here, also change it in the other method.

Reply:
I changed `update_` to `adjust_` in `MemoryPool` as it required the fewest changes and I do not have strong opinion on `adjust` vs `update`.